compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs

```diff
@@ -758,13 +758,37 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
         bounds: &mut Vec<(ty::Clause<'tcx>, Span)>,
         predicate_filter: PredicateFilter,
     ) -> GenericArgCountResult {
+        let tcx = self.tcx();
+
         // We use the *resolved* bound vars later instead of the HIR ones since the former
         // also include the bound vars of the overarching predicate if applicable.
         let hir::PolyTraitRef { bound_generic_params: _, modifiers, ref trait_ref, span } =
             *poly_trait_ref;
         let hir::TraitBoundModifiers { constness, polarity } = modifiers;
 
         let trait_def_id = trait_ref.trait_def_id().unwrap_or_else(|| FatalError.raise());
+
+        let (polarity, bounds) = match polarity {
+            rustc_ast::BoundPolarity::Positive
+                if tcx.is_lang_item(trait_def_id, hir::LangItem::PointeeSized) =>
+            {
+                // Skip `PointeeSized` bounds.
+                //
+                // `PointeeSized` is a "fake bound" insofar as anywhere a `PointeeSized` bound exists, there
+                // is actually the absence of any bounds. This avoids limitations around non-global where
+                // clauses being preferred over item bounds (where `PointeeSized` bounds would be
+                // proven) - which can result in errors when a `PointeeSized` supertrait/bound/predicate is
+                // added to some items.
+                (ty::PredicatePolarity::Positive, &mut Vec::new())
+            }
+            rustc_ast::BoundPolarity::Positive => (ty::PredicatePolarity::Positive, bounds),
+            rustc_ast::BoundPolarity::Negative(_) => (ty::PredicatePolarity::Negative, bounds),
+            // FIXME(fmease): This is super hacky! `Option<Polarity>` is also sad.
```

Comment:
(Yeah I should remove that FIXME, I forgot) I don't really like this structuring, very open to suggestions to address this FIXME.

**Bigger picture**: As you can see a few lines further down we used to duplicate the `lower_assoc_item_constraint` to catch bad bounds like `?Sized<Undefined = ()>`. This is obv annoying and not scalable (e.g., for `PointeeSized` we also need to validate the bound constness which is a lot more code I don't want to dupe / extract into a method (hmmm...)). So the new approach is to do as much normal lowering for these "fake bounds" but send everything into the drain (`&mut Vec::new()`).